File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/RootRouter.swift
import Foundation
import Observation
import WagerproofSharedKit

/// Top-level router that drives the root view's phase switch
/// (launching / auth / onboarding / main app). Mirrors RN's
/// `RootNavigator` + `OnboardingGuard` double-check in `app/_layout.tsx`.
@Observable
@MainActor
public final class RootRouter {
    public enum Phase: Equatable {
        case launching
        case unauthenticated
        case onboarding
        case ready
    }

    /// TEMPORARY: hard-bypasses the onboarding wizard so authenticated users
    /// land straight in the app. Added 2026-05-29 to unblock a creator who was
    /// stuck on an onboarding screen during testing. Flip back to `false` (or
    /// delete this flag and restore the plain ternary in `resolve`) to
    /// re-enable onboarding.
    public static let temporarilyDisableOnboarding = true

    /// Set by Secret Settings' "Reset Onboarding" so testers can re-enter the
    /// wizard even while `temporarilyDisableOnboarding` is active. In-memory
    /// only — cleared when onboarding completes; a relaunch falls back to the
    /// bypass, so this can never leak the wizard to real users.
    public private(set) var forceOnboardingForTesting = false

    public private(set) var phase: Phase = .launching

    /// The deep-link path captured before auth resolved. After auth completes,
    /// the root view replays it via `consumePendingDeepLink()`.
    public private(set) var pendingDeepLinkRoute: DeepLinkRoute?

    public init() {}

    public func resolve(authPhase: AuthStore.Phase, onboardingComplete: Bool) {
        switch authPhase {
        case .launching:
            phase = .launching
        case .unauthenticated:
            // Sign-out ends a forced test run — the next account shouldn't
            // inherit the wizard override.
            forceOnboardingForTesting = false
            phase = .unauthenticated
        case .authenticated:
            // Completing the wizard ends a forced test run so subsequent
            // resolves fall back to normal (bypassed) routing.
            if onboardingComplete { forceOnboardingForTesting = false }
            // TEMPORARY onboarding bypass — see `temporarilyDisableOnboarding`.
            // The bypass yields to `forceOnboardingForTesting` so the Secret
            // Settings reset can still run the wizard end-to-end.
            let bypass = Self.temporarilyDisableOnboarding && !forceOnboardingForTesting
            phase = (onboardingComplete || bypass) ? .ready : .onboarding
        }
    }

    /// Debug affordance backing Secret Settings' "Reset Onboarding": flips the
    /// router straight into `.onboarding`, overriding the temporary bypass.
    /// Callers must reset `OnboardingStore` first so the wizard starts at
    /// step 1 with a clean draft.
    public func forceOnboardingForTestingNow() {
        forceOnboardingForTesting = true
        phase = .onboarding
    }

    public func handle(deepLink url: URL) {
        guard let route = DeepLinkRoute(url: url) else { return }
        switch phase {
        case .ready:
            pendingDeepLinkRoute = route
        default:
            // Queue until auth resolves.
            pendingDeepLinkRoute = route
        }
    }

    /// Read+clear. Called by the root view's onChange(of: phase) handler once
    /// `.ready` is reached.
    public func consumePendingDeepLink() -> DeepLinkRoute? {
        defer { pendingDeepLinkRoute = nil }
        return pendingDeepLinkRoute
    }
}

/// Mirrors the deep-link map in `app/(drawer)/_layout.tsx`.
public enum DeepLinkRoute: Equatable, Sendable {
    case agents
    case outliers
    case feed
    case resetPassword

    public init?(url: URL) {
        guard url.scheme == "wagerproof" else { return nil }
        let host = url.host ?? url.pathComponents.dropFirst().first
        switch host {
        case "agents": self = .agents
        case "outliers": self = .outliers
        case "feed": self = .feed
        case "reset-password": self = .resetPassword
        default: self = .feed   // matches RN's default
        }
    }
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/RootRouter.kt
package com.wagerproof.core.stores

import android.net.Uri
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.shared.AppGroupKey

/**
 * Port of iOS `RootRouter.swift`. Top-level state machine behind the root
 * `when(phase)` composable switch (launching / auth / onboarding / main app).
 * Also queues deep links until auth resolves.
 */
@Stable
class RootRouter {

    enum class Phase { Launching, Unauthenticated, Onboarding, Ready }

    /**
     * Also set by "Reset Onboarding" — survives into [Phase.Ready] so the root
     * view can force the post-onboarding paywall for a Pro/admin tester.
     * Cleared via [clearTestPaywallOverride].
     */
    var testPaywallOverride by mutableStateOf(false); private set

    /** Swappable in local JVM tests so router tests never require Android prefs. */
    internal var testPaywallOverridePersistence: TestPaywallOverridePersistence =
        AppGroupTestPaywallOverridePersistence

    var phase by mutableStateOf(Phase.Launching); private set

    /** Deep-link route captured before auth resolved; replayed once [Phase.Ready]. */
    var pendingDeepLinkRoute by mutableStateOf<DeepLinkRoute?>(null); private set

    /** The ONLY phase mutator in normal flow. */
    fun resolve(authPhase: AuthStore.Phase, onboardingComplete: Boolean) {
        when (authPhase) {
            is AuthStore.Phase.Launching -> phase = Phase.Launching
            is AuthStore.Phase.Unauthenticated -> {
                // Sign-out ends a forced test run — the next account shouldn't inherit the override.
                clearTestPaywallOverride()
                phase = Phase.Unauthenticated
            }
            is AuthStore.Phase.Authenticated -> {
                phase = if (onboardingComplete) Phase.Ready else Phase.Onboarding
            }
        }
    }

    /**
     * Restore Secret Settings' pending replay before resolving the authenticated
     * root phase. A matching user keeps the override after process recreation;
     * a direct A -> B account switch clears it so the gate cannot cross accounts.
     */
    fun restoreTestPaywallOverride(userId: String) {
        val normalizedUserId = userId.trim().lowercase()
        val persistedUserId = testPaywallOverridePersistence.readUserId()
        testPaywallOverride = persistedUserId == normalizedUserId
        if (persistedUserId != null && !testPaywallOverride) {
            testPaywallOverridePersistence.clear()
        }
    }

    /** Backs Secret Settings' "Reset Onboarding". Caller must reset OnboardingStore first. */
    fun forceOnboardingForTestingNow(userId: String) {
        val normalizedUserId = userId.trim().lowercase()
        testPaywallOverridePersistence.writeUserId(normalizedUserId)
        testPaywallOverride = true
        phase = Phase.Onboarding
    }

    fun clearTestPaywallOverride() {
        testPaywallOverridePersistence.clear()
        testPaywallOverride = false
    }

    /** Parse + queue a deep link in ALL phases (queue-until-ready). */
    fun handle(uri: Uri) {
        val route = DeepLinkRoute.fromUri(uri) ?: return
        pendingDeepLinkRoute = route
    }

    /** Read+clear. Called by the root view's phase-change handler once [Phase.Ready]. */
    fun consumePendingDeepLink(): DeepLinkRoute? {
        val route = pendingDeepLinkRoute
        pendingDeepLinkRoute = null
        return route
    }
}

internal interface TestPaywallOverridePersistence {
    fun readUserId(): String?
    fun writeUserId(userId: String)
    fun clear()
}

private object AppGroupTestPaywallOverridePersistence : TestPaywallOverridePersistence {
    override fun readUserId(): String? = StorePrefs.appGroup
        .getString(AppGroupKey.TEST_PAYWALL_OVERRIDE_USER_ID, null)
        ?.trim()
        ?.lowercase()

    override fun writeUserId(userId: String) {
        StorePrefs.appGroup.edit()
            .putString(AppGroupKey.TEST_PAYWALL_OVERRIDE_USER_ID, userId)
            .apply()
    }

    override fun clear() {
        StorePrefs.appGroup.edit()
            .remove(AppGroupKey.TEST_PAYWALL_OVERRIDE_USER_ID)
            .apply()
    }
}

/** Mirrors the deep-link map in RN `app/(drawer)/_layout.tsx`. */
enum class DeepLinkRoute {
    Agents, Outliers, Feed, ResetPassword;

    companion object {
        fun fromUri(uri: Uri): DeepLinkRoute? {
            if (uri.scheme != "wagerproof") return null
            val host = uri.host ?: uri.pathSegments.firstOrNull()
            return when (host) {
                "agents" -> Agents
                "outliers" -> Outliers
                "feed" -> Feed
                "reset-password" -> ResetPassword
                else -> Feed // matches RN's default
            }
        }
    }
}
